### PR TITLE
Reduce allocations in Escape() and Unescape()

### PR DESCRIPTION
**Pull request by iSazonov**
# PR Summary

For scenario `dir c:\path\name*` we don't get big perf win (perhaps there is scenarios more sensetive) however before the fix  Escape() and Unescape() methods always allocate a temp buffer in heap. After the fix they use stack to get a temp buffer for paths < 256.

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [x] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/staging/reference/6/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): 
- **User-facing changes**
    - [x] Not Applicable
    - **OR**
    - [ ] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: 
- **Testing - New and feature**
    - [ ] N/A or can only be tested interactively
    - **OR**
    - [x] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)

### CODE REVIEW

File: src/System.Management.Automation/engine/regex.cs [hunk 1]
```diff
@@ -345,13 +353,18 @@ public static string Unescape(string pattern)
...
-            if (tempIndex > 0)
+            if (tempIndex == 0)
+            {
+                s = string.Empty;
+            }
+
+            if (tempIndex == pattern.Length)
             {
-                s = new string(temp, 0, tempIndex);
+                s = pattern;
```
Comment (TravisEz13) [inline src/System.Management.Automation/engine/regex.cs]: Do we have tests cases for regex escape?

File: src/System.Management.Automation/engine/regex.cs [hunk 2]
```diff
@@ -296,10 +302,12 @@ public static string Unescape(string pattern)
         {
             if (pattern == null)
             {
-                throw PSTraceSource.NewArgumentNullException("pattern");
+                throw PSTraceSource.NewArgumentNullException(nameof(pattern));
             }
 
-            char[] temp = new char[pattern.Length];
+            const int StackAllocThreshold = 256;
+            Span<char> temp = pattern.Length < StackAllocThreshold ? stackalloc char[StackAllocThreshold] : new char[pattern.Length];
```
Comment (daxian-dbw) [inline src/System.Management.Automation/engine/regex.cs]: > `? stackalloc char[StackAllocThreshold]`

Should be `stackalloc char[pattern.Length]`

File: src/System.Management.Automation/engine/regex.cs [hunk 3]
```diff
@@ -296,10 +302,12 @@ public static string Unescape(string pattern)
         {
             if (pattern == null)
             {
-                throw PSTraceSource.NewArgumentNullException("pattern");
+                throw PSTraceSource.NewArgumentNullException(nameof(pattern));
             }
 
-            char[] temp = new char[pattern.Length];
+            const int StackAllocThreshold = 256;
```
Comment (daxian-dbw) [inline src/System.Management.Automation/engine/regex.cs]: Should have only one const value for this.

File: src/System.Management.Automation/engine/regex.cs [hunk 4]
```diff
@@ -185,15 +185,16 @@ internal static string Escape(string pattern, char[] charsNotToEscape)
...
             }
 
             if (charsNotToEscape == null)
             {
-                throw PSTraceSource.NewArgumentNullException("charsNotToEscape");
+                throw PSTraceSource.NewArgumentNullException(nameof(charsNotToEscape));
             }
 
-            char[] temp = new char[pattern.Length * 2 + 1];
+            const int StackAllocThreshold = 256;
```
Comment (daxian-dbw) [inline src/System.Management.Automation/engine/regex.cs]: Why 256? How do you decide on the threshold?

File: src/System.Management.Automation/engine/regex.cs [hunk 5]
```diff
@@ -213,13 +214,18 @@ internal static string Escape(string pattern, char[] charsNotToEscape)
...
             string s = null;
 
-            if (tempIndex > 0)
+            if (tempIndex == 0)
             {
-                s = new string(temp, 0, tempIndex);
+                s = string.Empty;
+            }
+
+            if (tempIndex == pattern.Length)
```
Comment (daxian-dbw) [inline src/System.Management.Automation/engine/regex.cs]: `else if`?

File: src/System.Management.Automation/engine/regex.cs [hunk 6]
```diff
@@ -345,13 +353,18 @@ public static string Unescape(string pattern)
 
             string s = null;
 
-            if (tempIndex > 0)
+            if (tempIndex == 0)
+            {
+                s = string.Empty;
+            }
+
+            if (tempIndex == pattern.Length)
```
Comment (daxian-dbw) [inline src/System.Management.Automation/engine/regex.cs]: Same, `else if`?

**Review by daxian-dbw — CHANGES_REQUESTED** [review]

File: src/System.Management.Automation/engine/regex.cs (hunk 2 above)
Comment (iSazonov) [reply src/System.Management.Automation/engine/regex.cs]: Fixed.

File: src/System.Management.Automation/engine/regex.cs (hunk 3 above)
Comment (iSazonov) [reply src/System.Management.Automation/engine/regex.cs]: Fixed.

File: src/System.Management.Automation/engine/regex.cs (hunk 4 above)
Comment (iSazonov) [reply src/System.Management.Automation/engine/regex.cs]: Comment added: The size is less than MaxShortPath = 260

File: src/System.Management.Automation/engine/regex.cs (hunk 5 above)
Comment (iSazonov) [reply src/System.Management.Automation/engine/regex.cs]: Fixed.

File: src/System.Management.Automation/engine/regex.cs (hunk 6 above)
Comment (iSazonov) [reply src/System.Management.Automation/engine/regex.cs]: Fixed.

File: src/System.Management.Automation/engine/regex.cs (hunk 1 above)
Comment (iSazonov) [reply src/System.Management.Automation/engine/regex.cs]: Only indirectly - the API is used in engine and involved in paths, completions and other tests.
I open new tracking issue #10051 to add tests for WildcardPattern class.

File: src/System.Management.Automation/engine/regex.cs [hunk 7]
```diff
@@ -55,6 +55,12 @@ public sealed class WildcardPattern
         //
         private const char escapeChar = '`';
 
+        //
+        // Threshold for stack allocation.
+        // The size is less than MaxShortPath = 260.
+        //
+        private const int StackAllocThreshold = 256;
```
Comment (daxian-dbw) [inline src/System.Management.Automation/engine/regex.cs]: Why using `MaxShortPath` as the bar? We are using different thresholds at different places for `stackalloc` operations.

From SMA:
https://github.com/PowerShell/PowerShell/blob/master/src/System.Management.Automation/engine/Utils.cs#L208
https://github.com/PowerShell/PowerShell/blob/master/src/System.Management.Automation/FormatAndOutput/common/BaseOutputtingCommand.cs#L255

Maybe there are more in other dlls.

Comment (iSazonov) [reply src/System.Management.Automation/engine/regex.cs]: It depends on a context. The code can be involved for completions - it is very short strings, and for path globbing - common case is paths with lengths < MaxShortPath.
(We need a telemetry if we want see more real data.)

**Comment by iSazonov** [comment]
Rebased to resolve merge conflicts.

File: src/System.Management.Automation/engine/regex.cs (hunk 7 above)
Comment (daxian-dbw) [reply src/System.Management.Automation/engine/regex.cs]: corefx code doesn't have a single threshold either, so it should be fine to use different threshold in our code too, I guess.

File: src/System.Management.Automation/engine/regex.cs [hunk 8]
```diff
@@ -231,13 +235,17 @@ internal static string Escape(string pattern, char[] charsNotToEscape)
 
             string s = null;
 
-            if (tempIndex > 0)
+            if (tempIndex == 0)
```
Comment (daxian-dbw) [inline src/System.Management.Automation/engine/regex.cs]: This is basically `Pattern == string.Emtpy`. It should be checked before allocating the buffer, so we don't create a 1-char buffer for it.

File: src/System.Management.Automation/engine/regex.cs [hunk 9]
```diff
@@ -361,13 +370,17 @@ public static string Unescape(string pattern)
...
```
Comment (daxian-dbw) [inline src/System.Management.Automation/engine/regex.cs]: Same here.

File: src/System.Management.Automation/engine/regex.cs [hunk 10]
```diff
@@ -203,15 +207,15 @@ internal static string Escape(string pattern, char[] charsNotToEscape)
...
             }
 
             if (charsNotToEscape == null)
             {
-                throw PSTraceSource.NewArgumentNullException("charsNotToEscape");
+                throw PSTraceSource.NewArgumentNullException(nameof(charsNotToEscape));
             }
 
-            char[] temp = new char[pattern.Length * 2 + 1];
+            Span<char> temp = pattern.Length < StackAllocThreshold ? stackalloc char[pattern.Length * 2 + 1] : new char[pattern.Length * 2 + 1];
```
Comment (daxian-dbw) [inline src/System.Management.Automation/engine/regex.cs]: This is a bit misleading. `StackAllocThreshold` is used as the upper limit for the buffer to be allocated elsewhere. Here it's not the buffer size limit, but the pattern length limit.

**Review by daxian-dbw — CHANGES_REQUESTED** [review]

File: src/System.Management.Automation/engine/regex.cs (hunk 8 above)
Comment (iSazonov) [reply src/System.Management.Automation/engine/regex.cs]: Fixed.

File: src/System.Management.Automation/engine/regex.cs (hunk 9 above)
Comment (iSazonov) [reply src/System.Management.Automation/engine/regex.cs]: Fixed.

File: src/System.Management.Automation/engine/regex.cs (hunk 10 above)
Comment (iSazonov) [reply src/System.Management.Automation/engine/regex.cs]: Perhaps, but other options look even worse (I mean renaming and using 2 constants). On the other hand, in both cases we can consider this condition just as a boundary based on the length of the pattern.

File: src/System.Management.Automation/engine/regex.cs [hunk 11]
```diff
@@ -53,6 +53,10 @@ public sealed class WildcardPattern
         // char that escapes special chars
```
Comment (PoshChan) [inline src/System.Management.Automation/engine/regex.cs]: @iSazonov, your last commit had 1 failures in `PowerShell-CI-windows`
Resources strings in System.Management.Automation (was -ResGen used with Start-PSBuild).'Modules' should be an available type and the strings should be correct
```powershell
Expected strings to be the same, but they were different.
Expected length: 195
Actual length:   204
Strings differ at index 121.
Expected: '...psd1' and ...'
But was:  '...psd1', '.c...'
at <ScriptBlock>, D:\a\1\s\test\powershell\engine\ResourceValidation\TestRunner.ps1: line 64
64:                     NormalizeLineEnd($resourceStringToCheck) | Should -Be (NormalizeLineEnd($inresource.value))
```

Comment (PoshChan) [inline src/System.Management.Automation/engine/regex.cs]: @iSazonov, your last commit had 1 failures in `PowerShell-CI-windows`
Resources strings in System.Management.Automation (was -ResGen used with Start-PSBuild).'Modules' should be an available type and the strings should be correct
```powershell
Expected strings to be the same, but they were different.
Expected length: 195
Actual length:   204
Strings differ at index 121.
Expected: '...psd1' and ...'
But was:  '...psd1', '.c...'
at <ScriptBlock>, D:\a\1\s\test\powershell\engine\ResourceValidation\TestRunner.ps1: line 64
64:                     NormalizeLineEnd($resourceStringToCheck) | Should -Be (NormalizeLineEnd($inresource.value))
```

**Comment by iSazonov** [comment]
Codacy report is false positive.

**Comment by anmenaga** [comment]
@daxian-dbw Looks like your feedback was addressed; can you please update your review for this PR? Thank you.

**Review by daxian-dbw — APPROVED** [review]

**Comment by iSazonov** [comment]
@anmenaga Could you please merge?

**Pull request merged**